src/libstore/build/local-derivation-goal.cc

```diff
@@ -2936,16 +2936,20 @@ bool LocalDerivationGoal::isReadDesc(int fd)
 
 StorePath LocalDerivationGoal::makeFallbackPath(OutputNameView outputName)
 {
+    auto pathType = "rewrite:" + std::string(drvPath.to_string()) + ":name:" + std::string(outputName);
```

Comment:
This probably needs a comment, as this is very much a bogus "path type" --- in fact all these arguments are sort of spoofed and not what they usually are, the purpose is just to create store paths that shouldn't collide with any other store path.

Reply:
I don't think that's the right framing. This method is trying to use a wrong abstraction. Clearly `makeStorePath` was written for paths that are hashed in a certain way that isn't general enough anymore.

Wouldn't it be better to separate the responsibilities of constructing the path fingerprint (`s` in `makeStorePath`), and of actually doing the hashing? Any commonalities in fingerprint substrings could be considered coincidental and can be captured better by functions that return strings, rather than functions that return store paths. That's if we even want to bother with that. Wouldn't it be at least as clear if we just wrote the whole fingerprint concatenation every time?

Reply:
:scream: What have I done... I just wanted to understand how things work.

> Any commonalities in fingerprint substrings could be considered coincidental and can be captured better by functions that return strings, rather than functions that return store paths.

Yes, that. May need a slight refactoring worth 800 lines of diff though.

> Wouldn't it be at least as clear if we just wrote the whole fingerprint concatenation every time?

Surely not, because that's half of what I got stuck with trying to understand. The concatenation is not the problem. It's doing that seemingly ad hoc in place without explaining why these particular values are used. Maybe doing the concatenation inline is fine if it's commented properly and broken down to be readable? But then one may as well stuff it into a `struct`.

Reply:
```suggestion
    // This is a bogus path type constructed to ensure that it doesn't collide with any other store path 
    auto pathType = "rewrite:" + std::string(drvPath.to_string()) + ":name:" + std::string(outputName);
```

Reply:
> It's doing that seemingly ad hoc without explaining why these particular values are used

Just so we're clear, I think it's basically an arbitrary choice. The only real constraint is that the ways of generating store paths are disjoint. It could be `rewrite:` or it could be `valentin:`, it doesn't matter!

Reply:
I understand, but it's not evident from cursory reading of the code. That's why this change.

Reply:
> Maybe doing the concatenation inline is fine if it's commented properly and broken down to be readable?

I would think so yes, and I would expect it to need very little comment.

My point was basically that the code suggests we need abstraction while in the domain there is none. Bad abstractions are a distraction, and removing them can aid understanding.
This `makeStorePath` call seems to be such a case.

> But then one may as well stuff it into a `struct`.

We do have a general pattern where we start with _something_, then convert that to a fingerprint, and then convert that to a store path hash. Such a type would sit between _something_ and the conversion to a fingerprint (string). It seems that we wouldn't get much use out of the type then. It also seems that _something_ may need optional fields or even a variant. In such a case it's more natural to have a number of methods instead of an equivalent single use data type.